lectures/dynamic_programming/wald_friedman.md

```diff
@@ -360,162 +360,181 @@ Later, doing this will help us obey **the don't repeat yourself (DRY)** golden r
+clamp_belief(p) = clamp(float(p), BELIEF_FLOOR, one(Float64) - BELIEF_FLOOR)
 
-For every belief $p$, we can compute the difference between accepting a
-distribution and choosing to draw again.
+function gauss_jacobi_dist(F::Beta, N)
+    s, wj = FastGaussQuadrature.gaussjacobi(N, F.β - 1, F.α - 1)
+    x = (s .+ 1) ./ 2
+    C = 2.0^(-(F.α + F.β - 1.0)) / SpecialFunctions.beta(F.α, F.β)
+    return x, C .* wj
+end
 
-The solution $\alpha$, $\beta$ occurs at indifference points.
+function WFProblem(; d0 = Beta(1, 1), d1 = Beta(9, 9),
+                   L0 = 2.0, L1 = 2.0,
+                   c = 0.2,
+                   grid_size = 201,
+                   quad_order = 31)
+    belief_grid = collect(range(BELIEF_FLOOR,
+                                one(Float64) - BELIEF_FLOOR,
+                                length = grid_size))
+    nodes0, weights0 = gauss_jacobi_dist(d0, quad_order)
+    nodes1, weights1 = gauss_jacobi_dist(d1, quad_order)
+    return (; d0,
+            d1,
+            L0 = float(L0),
+            L1 = float(L1),
+            c = float(c),
+            belief_grid,
+            nodes0,
+            weights0,
+            nodes1,
+            weights1)
+end
 
-Define the cost function be the minimum of the pairwise differences in cost among
-the choices.
+function posterior_belief(problem, p, z)
+    (; d0, d1) = problem
+    num = p * pdf(d0, z)
+    den = num + (one(p) - p) * pdf(d1, z)
+    return den == 0 ? clamp_belief(p) : clamp_belief(num / den)
+end
 
-Then we can find the indifference points when the cost function is zero.
+function continuation_value(problem, p, vf)
+    (; c, nodes0, weights0, nodes1, weights1) = problem
 
-We can use any roots finding algorithm to solve for the solutions in the
-interval [0, 1].
+    loss(nodes, weights) =
+      dot(weights, vf.(posterior_belief.(Ref(problem), p, nodes)))
 
-Lastly, verify which indifference points correspond to the definition of a permanent
-transition between the accept and reject space for each choice.
+    return c + p * loss(nodes0, weights0) +
+           (one(p) - p) * loss(nodes1, weights1)
+end
+```
 
-Here's the code
+Next we solve a problem by applying value iteration to compute the value
+function and the associated decision rule.
 
 ```{code-cell} julia
-accept_x0(p, L0) = (one(p) - p) * L0
-accept_x1(p, L1) = p * L1
-function bayes_update(p, d0, d1)
-    p * pdf(d0, p) / pdf(MixtureModel([d0, d1], [p, one(p) - p]), p)
-end
-function draw_again(p, d0, d1, L0, L1, c, target)
-    candidate = 0.0
-    cost = 0.0
-    while candidate < target
-        p = bayes_update(p, d0, d1)
-        cost += c
-        candidate = min(accept_x0(p, L0), accept_x1(p, L1)) + cost
-        if candidate >= target
-            break
-        end
-        target = candidate
+function T(problem, v)
+    (; belief_grid, L0, L1) = problem
+    vf = extrapolate(interpolate((belief_grid,), v, Gridded(Linear())), Flat())
```

Comment:
Can you use `LinearInterpolation(belief_grid, v, extrapolation_bc = Flat())` or something like that to be consistent with the other uses of interpolation in the lectures?